File: kotlindemo/src/main/kotlin/com/example/kotlindemo/innerclass/InnerClass.java
package com.example.kotlindemo.innerclass;

/**
 * create by mozemin on 2018/11/9
 * desc:
 */
public class InnerClass {
    private int a;

    public class Inner {
        public void hello() {
            System.out.println(InnerClass.this.a);
        }
    }

    public static void main(String... args) {
        InnerClass innerClass = new InnerClass();
        View view = new View();
        view.setOnClicklistener(new OnClicklistener() {
            public void onClick() {

            }
        });
    }
}

File: kotlindemo/src/main/kotlin/com/example/kotlindemo/innerclass/InnerClass.kt
package com.example.kotlindemo.innerclass

/**
 * create by mozemin on 2018/11/9
 * desc:
 */

class Outter{
    val a:Int=10
    inner class Inner{// 非静态内部类
        val a :Int=5
        fun hello(){
            println("outter a:${this@Outter.a}")
        }
    }
    class Inner2{
        fun hello(){
            println("hello world!")
        }
    }
}
interface OnClicklistener{
    fun onClick()
}
class View{
    var onClicklistener:OnClicklistener?=null

}

fun main(args: Array<String>) {
    var inner = Outter().Inner()
    inner.hello()
    var inner2= Outter.Inner2()
    inner2.hello()

    val view=View()
    view.onClicklistener=object :OnClicklistener{
        override fun onClick() {
            TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
        }
    }

}